File: tools/article_swapper.js
// tools/article_swapper.js
const fs = require('fs');

function readFirst(paths){
  for (const p of paths){
    try { return fs.readFileSync(p, 'utf8'); } catch (_) {}
  }
  throw new Error('None of the candidate paths exist:\n' + paths.join('\n'));
}

// --- Candidate paths (kebab lowercase first, then existing mixed-case)
const CSS_PATHS   = ['templates/article-style-inline.css.html'];
const PDF_PATHS   = ['templates/snippets/pdf-button.html'];
const MAIN_PATHS  = ['templates/snippets/image-card-main.html',  'templates/snippets/image-card-Main.html'];
const LEFT_PATHS  = ['templates/snippets/image-card-left.html',  'templates/snippets/image-card-Left.html'];
const RIGHT_PATHS = ['templates/snippets/image-card-right.html', 'templates/snippets/image-card-Right.html'];
const BIO_PATHS   = ['templates/snippets/bio.html'];

// Load assets
const CSS_SENTINEL = '/* =========================\n       Tokens';
const CSS     = readFirst(CSS_PATHS);
const PDF_TPL = readFirst(PDF_PATHS);
const MAIN_T  = readFirst(MAIN_PATHS);
const LEFT_T  = readFirst(LEFT_PATHS);
const RIGHT_T = readFirst(RIGHT_PATHS);
const BIO_T   = readFirst(BIO_PATHS);

// === A) CSS injection (idempotent)
function injectCss(html){
  if (html.includes(CSS_SENTINEL)) return html;
  if (/<\/head>/i.test(html)) return html.replace(/<\/head>/i, CSS + '\n</head>');
  if (/<body[^>]*>/i.test(html)) return html.replace(/<body[^>]*>/i, (m)=> m + '\n' + CSS + '\n');
  return CSS + '\n' + html;
}

// Expand [photo ...] tokens
function expandPhotoTokens(html){
  return html.replace(/\[photo\s+(left|right|main)\]/gi, (_, pos)=>{
    switch (pos.toLowerCase()){
      case 'left':  return LEFT_T;
      case 'right': return RIGHT_T;
      default:      return MAIN_T;
    }
  });
}

// Ensure <header class="article-header">
function ensureHeaderClass(openTag){
  if (/class=/i.test(openTag)){
    return openTag.replace(/class=(["'])(.*?)\1/i, (m,q,cls)=>{
      return /\barticle-header\b/i.test(cls) ? m : `class=${q}${cls} article-header${q}`;
    });
  }
  return openTag.replace(/>$/, ' class="article-header">');
}

// B) Header normalization
function normalizeHeader(html){
  return html.replace(/(<header[^>]*>)([\s\S]*?)(<\/header>)/i, (whole, open, inner, close)=>{
    let body = inner;
    open = ensureHeaderClass(open);

    // Author -> h3.author
    body = body.replace(
      /<p[^>]*>\s*(?:<strong>)?\s*By\s*[:—-]?\s+([^<]+?)(?:<\/strong>)?\s*<\/p>/i,
      (_m, name)=> `<h3 class="author">By ${name.trim()}</h3>`
    );

    // Organization -> first p after author → h4.org
    body = body.replace(
      /(<h3[^>]*class=["'][^"']*\bauthor\b[^"']*["'][^>]*>[\s\S]*?<\/h3>)([\s\S]*?)(<p(?![^>]*class=["'][^"']*\bpubdate\b)[^>]*>[\s\S]*?<\/p>)/i,
      (_m, author, mid, p) => `${author}${mid}${p.replace(/^<p/i,'<h4 class="org"').replace(/<\/p>$/i,'</h4>')}`
    );

    // Pubdate -> p.pubdate (or placeholder)
    if (!/<p[^>]*class=["'][^"']*\bpubdate\b/i.test(body)){
      body = body.replace(
        /<p>([A-Za-z]{3,}\s+\d{1,2},\s+\d{4})<\/p>/,
        '<p class="pubdate">$1</p>'
      );
      if (!/<p[^>]*class=["'][^"']*\bpubdate\b/i.test(body)){
        body += '\n<p class="pubdate">(date goes here)</p>';
      }
    }

    // Marker to anchor later inserts
    body = body.replace(/(<p[^>]*class=["'][^"']*\bpubdate\b[^"']*["'][^>]*>[\s\S]*?<\/p>)/i, '$1\n<!--__AFTER_PUBDATE__-->'
);
    return `${open}${body}${close}`;
  });
}

// C) PDF placement (dedupe-safe)
function insertPdfAfterDate(html){
  const m = html.match(/(<header[^>]*>)([\s\S]*?)(<\/header>)/i);
  if (!m) return html;
  const [whole, open, inner, close] = m;

  if (/(pdficon_small\.png|Download the PDF)/i.test(inner)){
    const tagged = inner.replace(
      /(<p[^>]*>[\s\S]*?(?:pdficon_small\.png|Download the PDF)[\s\S]*?<\/p>)/i,
      '$1\n<!--__AFTER_PDF__-->'
    );
    return html.replace(whole, `${open}${tagged}${close}`);
  }
  if (html.includes('<!--__AFTER_PUBDATE__-->')){
    return html.replace('<!--__AFTER_PUBDATE__-->', `${PDF_TPL}\n<!--__AFTER_PDF__-->`);
  }
  return html.replace(/<\/header>/i, `${PDF_TPL}\n<!--__AFTER_PDF__-->\n</header>`);
}

// D) Main image hoist / placeholder
function hoistMainAfterPdf(html){
  if (!html.includes('<!--__AFTER_PDF__-->')) return html;
  const mainRe = /<figure[^>]*class=["'][^"']*\bimage-card\b[^"']*\bMain\b[^"']*["'][^>]*>[\s\S]*?<\/figure>/i;
  const found = mainRe.exec(html);
  if (found){
    const block = found[0];
    const without = html.slice(0, found.index) + html.slice(found.index + block.length);
    return without.replace('<!--__AFTER_PDF__-->', `${block}\n<!--__AFTER_PDF__-->`);
  }
  return html.replace('<!--__AFTER_PDF__-->', `${MAIN_T}\n<!--__AFTER_PDF__-->`);
}

// E) References normalization
function normalizeReferences(html){
  const refsH3 = /<h3[^>]*>\s*References\s*<\/h3>/i;
  if (!refsH3.test(html)) return html;

  return html.replace(
    /(<h3[^>]*>\s*References\s*<\/h3>)([\s\S]*?)(?=(<\/section>|<h2|<h3|<\/article>|$))/i,
    (whole, h3, block, tail) => {
      let out = block;

      // Ensure <p class="reference">
      out = out.replace(/<p(?![^>]*class=)[^>]*>/gi, '<p class="reference">');
      out = out.replace(/<p\s+class=(["'])(?![^"']*\breference\b)[^"']*\1/gi,
        m => m.replace(/class=(['"])[^'"]*\1/, 'class="reference"'));

      // <span> -> <em>
      out = out.replace(/<\/?span>/gi, m => m[1] === '/' ? '</em>' : '<em>');

      // Add link attrs
      out = out.replace(/<a([^>]*?)href=(["'])(.*?)\2([^>]*)>/gi, (_m, pre, q, url, post) => {
        const meta = (pre+post).toLowerCase();
        const target = meta.includes('target=') ? '' : ' target="_blank"';
        const rel    = meta.includes('rel=')    ? '' : ' rel="noopener"';
        const onclick= meta.includes('onclick=')? '' : ' onclick="_gaq.push([\'_trackEvent\',\'Notes Link\',\'Click\', this.href]);"';
        return `<a${pre}href="${url}"${post}${target}${rel}${onclick}>`;
      });

      return h3 + out + (tail ?? '');
    }
  );
}

// F) Bio after the last reference
function insertBioAfterReferences(html){
  if (/<div[^>]*class=["'][^"']*\bbio-card\b/i.test(html)) return html;

  // Try contiguous reference block
  const block = new RegExp(
    '(<h3[^>]*>\\s*References\\s*<\\/h3>)' +
    '((?:\\s*<p\\s+class="reference"[^>]*>[\\s\\S]*?<\\/p>)+)',
    'i'
  );
  const m = html.match(block);
  if (m){
    return html.replace(block, `${m[1]}${m[2]}\n${BIO_T}`);
  }

  // If only the heading exists (no normalized ref paragraphs yet), insert after heading
  if (/<h3[^>]*>\s*References\s*<\/h3>/i.test(html)){
    return html.replace(/(<h3[^>]*>\s*References\s*<\/h3>)/i, `$1\n${BIO_T}\n`);
  }

  // No references at all → append near end
  return html.replace(/<\/article>|<\/body>/i, (t)=> `${BIO_T}\n${t}`);
}

// G) Final cleanup
function finalCleanup(html){
  return html
    .replace(/<!--__AFTER_PUBDATE__-->/g, '')
    .replace(/<!--__AFTER_PDF__-->/g, '')
    .replace(/\n{3,}/g, '\n\n');
}

// === Runner
if (require.main === module){
  const [,, inFile, outFile] = process.argv;
  if (!inFile || !outFile){
    console.error('Usage: node tools/article_swapper.js input.html output.html');
    process.exit(1);
  }
  let out = fs.readFileSync(inFile, 'utf8');
  out = injectCss(out);
  out = expandPhotoTokens(out);
  out = normalizeHeader(out);
  out = insertPdfAfterDate(out);
  out = hoistMainAfterPdf(out);
  out = normalizeReferences(out);
  out = insertBioAfterReferences(out);
  out = finalCleanup(out);
  fs.writeFileSync(outFile, out, 'utf8');
  console.log(`Wrote ${outFile}`);
}

File: tools/article_swapper.py
from pathlib import Path
import re
import sys

CSS_PATH  = Path('templates/article-style-inline.css.html')
MAIN_TPL  = Path('templates/snippets/image-card-Main.html')
LEFT_TPL  = Path('templates/snippets/image-card-Left.html')
RIGHT_TPL = Path('templates/snippets/image-card-Right.html')
BIO_TPL   = Path('templates/snippets/bio.html')
PDF_TPL   = Path('templates/snippets/pdf-button.html')

CSS_SENTINEL = '/* =========================\n       Tokens'

def inject_css(html: str, css: str) -> str:
    if CSS_SENTINEL in html:
        return html
    if re.search(r'</head>', html, flags=re.I):
        return re.sub(r'</head>', css + '\n</head>', html, count=1, flags=re.I)
    if re.search(r'<body[^>]*>', html, flags=re.I):
        return re.sub(r'(<body[^>]*>)', r'\1\n' + css + '\n', html, count=1, flags=re.I)
    return css + '\n' + html

def expand_photo_tokens(html: str, main_tpl: str, left_tpl: str, right_tpl: str) -> str:
    def repl(m):
        pos = m.group(1).lower()
        return {'left': left_tpl, 'right': right_tpl, 'main': main_tpl}[pos]
    return re.sub(r'\[photo\s+(left|right|main)\]', repl, html, flags=re.I)

def ensure_class_on_header(open_tag: str) -> str:
    if re.search(r'class=', open_tag, flags=re.I):
        def repl(m):
            cls = m.group(2)
            return f'class="{cls} article-header"' if 'article-header' not in cls.lower() else m.group(0)
        return re.sub(r'class=(["\'])(.*?)\1', repl, open_tag, count=1, flags=re.I)
    return open_tag[:-1] + ' class="article-header">'

def normalize_header(html: str) -> str:
    m = re.search(r'(<header[^>]*>)([\s\S]*?)(</header>)', html, flags=re.I)
    if not m:
        return html
    open_tag, inner, close_tag = m.groups()
    open_tag = ensure_class_on_header(open_tag)
    body = inner

    # Author
    body = re.sub(r'<p[^>]*>\s*(?:<strong>)?\s*By\s*[:—-]?\s+([^<]+?)(?:</strong>)?\s*</p>',
                  lambda mm: f'<h3 class="author">By {mm.group(1).strip()}</h3>',
                  body, count=1, flags=re.I)

    # Org (first remaining p after author)
    body = re.sub(r'(<h3[^>]*class=["\'][^"\']*\bauthor\b[^"\']*["\'][^>]*>[\s\S]*?</h3>)([\s\S]*?)(<p(?![^>]*class=["\'][^"\']*pubdate)[\s\S]*?</p>)',
                  lambda mm: mm.group(1) + mm.group(2) + mm.group(3).replace('<p','<h4 class="org"').replace('</p>','</h4>'),
                  body, count=1, flags=re.I)

    # Pubdate
    if not re.search(r'<p[^>]*class=["\'][^"\']*\bpubdate\b', body, flags=re.I):
        body = re.sub(r'<p>([A-Za-z]{3,}\s+\d{1,2},\s+\d{4})</p>',
                      r'<p class="pubdate">\1</p>', body, count=1, flags=re.I)
        if not re.search(r'<p[^>]*class=["\'][^"\']*\bpubdate\b', body, flags=re.I):
            body += '\n<p class="pubdate">(date goes here)</p>'

    # Marker after pubdate
    body = re.sub(r'(<p[^>]*class=["\'][^"\']*\bpubdate\b[^"\']*["\'][^>]*>[\s\S]*?</p>)',
                  r'\1\n<!--__AFTER_PUBDATE__-->', body, count=1, flags=re.I)

    new_header = f'{open_tag}{body}{close_tag}'
    return html[:m.start()] + new_header + html[m.end():]

def insert_pdf_after_date(html: str, pdf_tpl: str) -> str:
    hm = re.search(r'(<header[^>]*>)([\s\S]*?)(</header>)', html, flags=re.I)
    if not hm:
        return html
    open_tag, inner, close_tag = hm.groups()

    if re.search(r'(pdficon_small\.png|Download the PDF)', inner, flags=re.I):
        inner = re.sub(r'(<p[^>]*>[\s\S]*?(?:pdficon_small\.png|Download the PDF)[\s\S]*?</p>)',
                       r'\1\n<!--__AFTER_PDF__-->', inner, count=1, flags=re.I)
        return html[:hm.start()] + open_tag + inner + close_tag + html[hm.end():]

    if '<!--__AFTER_PUBDATE__-->' in html:
        return html.replace('<!--__AFTER_PUBDATE__-->', pdf_tpl + '\n<!--__AFTER_PDF__-->')

    return re.sub(r'</header>', pdf_tpl + '\n<!--__AFTER_PDF__-->\n</header>', html, count=1, flags=re.I)

def hoist_main_after_pdf(html: str, main_tpl: str) -> str:
    if '<!--__AFTER_PDF__-->' not in html:
        return html
    m = re.search(r'<figure[^>]*class=["\'][^"\']*\bimage-card\b[^"\']*\bMain\b[^"\']*["\'][^>]*>[\s\S]*?</figure>', html, flags=re.I)
    if m:
        block = m.group(0)
        without = html[:m.start()] + html[m.end():]
        return without.replace('<!--__AFTER_PDF__-->', block + '\n<!--__AFTER_PDF__-->')
    return html.replace('<!--__AFTER_PDF__-->', main_tpl + '\n<!--__AFTER_PDF__-->')

def normalize_references(html: str) -> str:
    refs_head = re.search(r'(<h3[^>]*>\s*References\s*</h3>)', html, flags=re.I)
    if not refs_head:
        return html
    start = refs_head.end()
    tail_match = re.search(r'(</section>|<h2|<h3|</article>|</div>\s*</div>\s*</div>|$)', html[start:], flags=re.I)
    end = start + (tail_match.start() if tail_match else 0)
    block = html[start:end]

    def fix_p(m):
        tag = m.group(0)
        if re.search(r'class=["\'][^"\']*\breference\b', tag, flags=re.I):
            return tag
        if not re.search(r'class=', tag, flags=re.I):
            return tag.replace('<p','<p class="reference"', 1)
        return re.sub(r'class=["\'][^"\']*["\']', 'class="reference"', tag, flags=re.I)
    block = re.sub(r'<p[^>]*>', fix_p, block, flags=re.I)

    block = re.sub(r'</?span>', lambda mm: '</em>' if mm.group(0).startswith('</') else '<em>', block, flags=re.I)

    def link_repl(mm):
        pre, quote, url, post = mm.group(1), mm.group(2), mm.group(3), mm.group(4)
        meta = (pre + post).lower()
        if 'target=' not in meta:
            post += ' target="_blank"'
        if 'rel=' not in meta:
            post += ' rel="noopener"'
        if 'onclick=' not in meta:
            post += ' onclick="_gaq.push([\'_trackEvent\',\'Notes Link\',\'Click\', this.href]);"'
        return f'<a{pre}href="{url}"{post}>'
    block = re.sub(r'<a([^>]*?)href=(["\'])(.*?)\2([^>]*)>', link_repl, block, flags=re.I)

    return html[:start] + block + html[end:]

def insert_bio_if_missing(html: str, bio_tpl: str) -> str:
    if re.search(r'<div[^>]*class=["\'][^"\']*\bbio-card\b', html, flags=re.I):
        return html
    ref_h3 = re.search(r'<h3[^>]*>\s*References\s*</h3>', html, flags=re.I)
    if not ref_h3:
        return re.sub(r'</article>|</div>\s*</div>\s*</div>\s*</body>', bio_tpl + '\n\\g<0>', html, flags=re.I)
    idx = ref_h3.start()
    after_refs = html[idx:]
    close_sec = re.search(r'</section>', after_refs, flags=re.I)
    if close_sec:
        insert_at = idx + close_sec.end()
        return html[:insert_at] + '\n' + bio_tpl + '\n' + html[insert_at:]
    return html.replace(ref_h3.group(0), ref_h3.group(0) + '\n' + bio_tpl + '\n', 1)

def final_cleanup(html: str) -> str:
    html = html.replace('<!--__AFTER_PUBDATE__-->', '').replace('<!--__AFTER_PDF__-->', '')
    html = re.sub(r'\n{3,}', '\n\n', html)
    return html

def main():
    if len(sys.argv) < 3:
        print('Usage: python tools/article_swapper.py input.html output.html')
        sys.exit(1)
    in_file, out_file = Path(sys.argv[1]), Path(sys.argv[2])

    css     = CSS_PATH.read_text(encoding='utf-8')
    html    = in_file.read_text(encoding='utf-8')
    main_t  = MAIN_TPL.read_text(encoding='utf-8')
    left_t  = LEFT_TPL.read_text(encoding='utf-8')
    right_t = RIGHT_TPL.read_text(encoding='utf-8')
    bio_t   = BIO_TPL.read_text(encoding='utf-8')
    pdf_t   = PDF_TPL.read_text(encoding='utf-8')

    out = inject_css(html, css)
    out = expand_photo_tokens(out, main_t, left_t, right_t)
    out = normalize_header(out)
    out = insert_pdf_after_date(out, pdf_t)
    out = hoist_main_after_pdf(out, main_t)
    out = normalize_references(out)
    out = insert_bio_if_missing(out, bio_t)
    out = final_cleanup(out)

    Path(out_file).write_text(out, encoding='utf-8')
    print(f'Wrote {out_file}')

if __name__ == '__main__':
    main()
